core/unit_test/TestTeamReductionScan.hpp

```diff
@@ -97,5 +97,51 @@ TEST(TEST_CATEGORY, team_double_reduce) {
   }
 }
 
+template <typename ExecutionSpace>
+struct DummyTeamReductionFunctor {
+  using TeamPolicy     = Kokkos::TeamPolicy<ExecutionSpace>;
+  using TeamHandleType = typename TeamPolicy::member_type;
+
+  KOKKOS_FUNCTION void operator()(const TeamHandleType&, double&) const {}
+};
+
+template <typename ExecutionSpace>
+void test_team_parallel_reduce(const int num_loop_size) {
+  using TeamPolicy = Kokkos::TeamPolicy<ExecutionSpace>;
+
+  using ReducerType = Kokkos::Sum<double>;
+  double result     = 10.;
+  ReducerType reducer(result);
```

Comment:
I initialize it so that I can see that it's actually set to zero in the reduction (and not by chance uninitialized zero).
I need the reducer to trigger the bug. Otherwise `needs_device_set` would not be `true`.

Reply:
But that what it would default to if you passed a scalar

Reply:
https://github.com/kokkos/kokkos/blob/ecbfc2f7d3902061d3e60ed7861845925ad8d48e/core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp#L866-L874
Using a reducer I get both `has_init_member_function` and `has_reducer`. If I only provide a scalar `need_device_set` is false.

```
init: 1
final: 0
host accessible: 1
has reducer: 1
```
for the reducer case for the different parts of `need_device_set`
```
init: 0
final: 0
host accessible: 1
has reducer: 0 
```
for the scalar case.